linker.ts

```diff
@@ -1,72 +1,132 @@
+ *
+ * This method provides a simple helper for linking.
+ *
+ * @param {string} bytecode
+ * @param {string} libraries
+ * @returns {string} bytecode
+ */
+function linkBytecode (bytecode, libraries) {
   assert(typeof bytecode === 'string');
   assert(typeof libraries === 'object');
+
   // NOTE: for backwards compatibility support old compiler which didn't use file names
   const librariesComplete = {};
-  for (const libraryName in libraries) {
-    if (typeof libraries[libraryName] === 'object') {
-      // API compatible with the standard JSON i/o
-      for (const lib in libraries[libraryName]) {
-        librariesComplete[lib] = libraries[libraryName][lib];
-        librariesComplete[libraryName + ':' + lib] = libraries[libraryName][lib];
-      }
-    } else {
-      // backwards compatible API for early solc-js versions
-      const parsed = libraryName.match(/^([^:]+):(.+)$/);
-      if (parsed) {
-        librariesComplete[parsed[2]] = libraries[libraryName];
+
+  for (const [libraryName, library] of Object.entries(libraries)) {
+    if (isNil(library)) continue;
+
+    // API compatible with the standard JSON i/o
+    if (isObject(library)) {
+      for (const [key, value] of Object.entries(library)) {
+        librariesComplete[key] = value;
+        librariesComplete[`${libraryName}:${key}`] = value;
       }
-      librariesComplete[libraryName] = libraries[libraryName];
+
+      continue;
     }
+
+    // backwards compatible API for early solc-js versions
+    const parsed = libraryName.match(/^([^:]+):(.+)$/);
+
+    if (!isNil(parsed)) librariesComplete[parsed[2]] = library;
+    librariesComplete[libraryName] = library;
   }
 
   for (const libraryName in librariesComplete) {
     let hexAddress = librariesComplete[libraryName];
-    if (hexAddress.slice(0, 2) !== '0x' || hexAddress.length > 42) {
-      throw new Error('Invalid address specified for ' + libraryName);
+
+    if (!hexAddress.startsWith('0x') || hexAddress.length > 42) {
+      throw new Error(`Invalid address specified for ${libraryName}`);
     }
+
     // remove 0x prefix
     hexAddress = hexAddress.slice(2);
     hexAddress = Array(40 - hexAddress.length + 1).join('0') + hexAddress;
 
-    // Support old (library name) and new (hash of library name)
-    // placeholders.
-    const replace = function (name) {
-      // truncate to 37 characters
-      const truncatedName = name.slice(0, 36);
-      const libLabel = '__' + truncatedName + Array(37 - truncatedName.length).join('_') + '__';
-      while (bytecode.indexOf(libLabel) >= 0) {
-        bytecode = bytecode.replace(libLabel, hexAddress);
-      }
-    };
-
-    replace(libraryName);
-    replace(libraryHashPlaceholder(libraryName));
+    // replace all library names with a hex address of all zeros of length
+    // (40 - len(hexAddress) + 1) + the org hex address without the prefix.
+    //
+    // and then again for the hash version of hte library name.
```

Comment:
I don't fully understand this comment. What is "org address" in this context?

Anyway, this sounds like something that should really be explained by the docstring and not here. See my next suggestion.